% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/filter_host.R
\name{remove_matches}
\alias{remove_matches}
\title{Helper function to remove reads matched to filter libraries}
\usage{
remove_matches(reads_bam, read_names, name_out, YS, threads, aligner)
}
\arguments{
\item{reads_bam}{The name of a merged, sorted .bam file that has previously
been aligned to a reference library. Likely, the output from running an
instance of \code{align_target()}.}

\item{read_names}{A \code{list} of target query names from \code{reads_bam}
that have also aligned to a filter reference library. Each \code{list}
element should be a vector of read names.}

\item{name_out}{The name of the .bam file that to which the filtered alignments
will be written.}

\item{YS}{yieldSize, an integer. The number of alignments to be read in from
the bam file at once. Default is 1000000.}

\item{threads}{The number of threads to be used in filtering the bam file.}

\item{aligner}{The aligner which was used to create the bam file.}
}
\value{
The name of a filtered, sorted .bam file written to the user's
current working directory.
}
\description{
Within the \code{filter_host()} function, we align our sequencing sample to all
filter libraries of interest. The \code{remove_matches()} function allows
for removal of any target reads that are also aligned to filter libraries.
}
\details{
This function is not intended for direct use.
}
\examples{

#readPath <- system.file("extdata", "subread_target.bam",
#                        package = "MetaScope")

## Assume that the first 10 query names aligned to first filter library
## And another 10 aligned to second filter library
# qnames <- Rsamtools::scanBam(readPath)[[1]]$qname
# read_names <- list(qnames[1:10], qnames[30:40])
# out <- "subread_target.filtered.bam"

# remove_matches(readPath, read_names, out)

}
